### PR TITLE
[sgd] Refactor PyTorch SGD Documentation.

### PR DESCRIPTION
**Pull request by richardliaw**
## Why are these changes needed?
Improving RaySGD documentation.

image
image

TODO:
 - [x] Fix comments, add examples
 - [x] Test on cluster once more.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: doc/source/raysgd/raysgd_example.rst [hunk 1]
```diff
@@ -0,0 +1,15 @@
+RaySGD Examples
+===============
+
+Here are some examples of using RaySGD for training PyTorch models.
+
```
Comment (richardliaw) [inline doc/source/raysgd/raysgd_example.rst]: TODO

File: python/ray/experimental/sgd/pytorch/pytorch_trainer.py [hunk 2]
```diff
@@ -45,13 +46,16 @@ def __init__(self,
         Args:
             model_creator (dict -> torch.nn.Module): creates the model
                 using the config.
-            data_creator (int, dict -> DataLoader, DataLoader): Function that
-                takes in (batch_size, config) and returns two Torch DataLoader
+            data_creator (int, dict -> Dataset, Dataset): Function that
+                takes in (batch_size, config) and returns two Torch Dataset
```
Comment (richardliaw) [inline python/ray/experimental/sgd/pytorch/pytorch_trainer.py]: one or two

File: python/ray/experimental/sgd/pytorch/pytorch_trainer.py [hunk 3]
```diff
@@ -45,13 +46,16 @@ def __init__(self,
...
                 objects.
             optimizer_creator (torch.nn.Module, dict -> optimizer):
                 creates the loss and optimizer using the model and the config.
-            loss_creator (dict -> loss): Creates the loss function/criterion
-                using the config.
+            loss_creator (dict -> loss or torch.nn.*Loss): A creator function
+                for the loss function. This can be either a function that
+                takes in the provided config for customization or a subclass
+                of ``torch.nn.modules.loss._Loss``, which is most Pytorch
+                loss classes. For example, ``loss_creator=torch.nn.BCELoss``.
```
Comment (richardliaw) [inline python/ray/experimental/sgd/pytorch/pytorch_trainer.py]: Add also `dataloader_config`.

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 4]
```diff
@@ -3,114 +3,343 @@ RaySGD Pytorch
...
+Under the hood, ``PytorchTrainer`` will create *replicas* of your model (controlled by ``num_replicas``) which are each managed by a Ray actor.
 
-Wrap your training with this:
+.. image:: raysgd-actors.svg
+    :align: center
+
+Setting up training
+-------------------
+
+The ``PyTorchTrainer`` can be constructed with functions that construct components of the training script. Specifically, it needs constructors for the Model, Data, Optimizer, and Loss. This is so that we can create replicated copies of the neural network/optimizer across different devices and machines.
```
Comment (richardliaw) [inline doc/source/raysgd/raysgd_pytorch.rst]: In this paragraph, we should also mention the cases where we take in different types for each argument.

File: python/ray/experimental/sgd/pytorch/pytorch_runner.py [hunk 5]
```diff
@@ -24,12 +24,13 @@ def __init__(self,
                  train_function=None,
                  validation_function=None,
                  config=None,
+                 dataloader_config=None,
```
Comment (richardliaw) [inline python/ray/experimental/sgd/pytorch/pytorch_runner.py]: Add to docstring

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 6]
```diff
@@ -3,114 +3,390 @@ RaySGD Pytorch
...
-PyTorchTrainer Example
-----------------------
+    trainer.train(max_retries=N)
 
-Below is an example of using Ray's PyTorchTrainer. Under the hood, ``PytorchTrainer`` will create *replicas* of your model (controlled by ``num_replicas``) which are each managed by a worker.
 
-.. literalinclude:: ../../../python/ray/experimental/sgd/examples/train_example.py
-   :language: python
-   :start-after: __torch_train_example__
+During each ``train`` method, each parallel worker iterates through the dataset, synchronizing gradients and parameters at each batch. These synchronization primitives can hang when one or more of the parallel workers becomes unresponsive (i.e., when a node is lost). To address this, we've implemented the following protocol.
```
Comment (deanwampler) [inline doc/source/raysgd/raysgd_pytorch.rst]: Do you want to discuss how this synchronization is done. For example, are gradients and parameters averaged together in some way?

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 7]
```diff
@@ -3,114 +3,390 @@ RaySGD Pytorch
...
-.. autoclass:: ray.experimental.sgd.pytorch.PyTorchTrainer
-    :members:
+    def model_creator(config):
+        netD = Discriminator()
+        netD.apply(weights_init)
 
-    .. automethod:: __init__
+        netG = Generator()
+        netG.apply(weights_init)
+        return netD, netG
```
Comment (deanwampler) [inline doc/source/raysgd/raysgd_pytorch.rst]: I wonder if we should return lists when more than 1 is returned? Or should it always return a list for uniformity, even when just 1 is returned? Similarly, should any function that accepts model(s) as an argument always take a list?

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 8]
```diff
@@ -3,114 +3,390 @@ RaySGD Pytorch
...
+        data_creator,
+        optimizer_creator,
+        nn.BCELoss)
+
+.. _raysgd-custom-training:
+
+Custom Training and Validation Functions
+----------------------------------------
+
+``PyTorchTrainer`` allows you to run a custom training and validation in parallel on each worker, providing a flexibility similar to using PyTorch natively. This is done via the ``train_function`` and ``validation_function`` parameters.
```
Comment (deanwampler) [inline doc/source/raysgd/raysgd_pytorch.rst]: "custom training and validation step in ..." ? (add "step")

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 9]
```diff
@@ -3,114 +3,390 @@ RaySGD Pytorch
...
+
+.. code-block:: python
+
+    def train(models, dataloader, criterion, optimizers, config):
+        """A custom training function.
+
+        Args:
+            models: Output of the model_creator passed into PyTorchTrainer.
+            data_loader: A dataloader wrapping the training dataset created by the ``data_creator`` passed into PyTorchTrainer.
+            criterion: The instantiation of the ``loss_creator``.
```
Comment (deanwampler) [inline doc/source/raysgd/raysgd_pytorch.rst]: Why now use "loss" instead of "criterion" for consistency?

**Review by deanwampler — APPROVED** [review]
Looks great. I left some comments for minor text suggestions and other comments that are more about the API itself.

**Comment by richardliaw** [comment]
jenkins test tune

**Comment by richardliaw** [comment]
jenkins test tune

File: doc/source/raysgd/raysgd.rst [hunk 10]
```diff
@@ -1,24 +1,65 @@
 RaySGD: Distributed Deep Learning
 =================================
 
+.. tip:: Get in touch with us if you're using/considering using `RaySGD <https://forms.gle/26EMwdahdgm7Lscy9>`_!
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/raysgd/raysgd.rst]:
```suggestion
.. tip:: Get in touch with us if you're using or considering using `RaySGD <https://forms.gle/26EMwdahdgm7Lscy9>`_!
```

File: doc/source/raysgd/raysgd.rst [hunk 11]
```diff
@@ -1,24 +1,65 @@
...
-.. tip:: Help us make RaySGD better; take this 1 minute `User Survey <https://forms.gle/26EMwdahdgm7Lscy9>`_!
+RaySGD is a lightweight library for distributed deep learning, providing thin wrappers around PyTorch and TensorFlow native modules for data parallel training.
 
 The main features are:
 
-  - Ease of use: Scale Pytorch's native ``DistributedDataParallel`` and TensorFlow's ``tf.distribute.MirroredStrategy`` without needing to monitor individual nodes.
-  - Composibility: RaySGD is built on top of the Ray Actor API, enabling seamless integration with existing Ray applications such as RLlib, Tune, and Ray.Serve.
-  - Scale up and down: Start on single CPU. Scale up to multi-node, multi-gpu by changing 2 lines of code.
+  - **Ease of use**: Scale Pytorch's native ``DistributedDataParallel`` and TensorFlow's ``tf.distribute.MirroredStrategy`` without needing to monitor individual nodes.
+  - **Composibility**: RaySGD is built on top of the Ray Actor API, enabling seamless integration with existing Ray applications such as RLlib, Tune, and Ray.Serve.
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/raysgd/raysgd.rst]: english is not my strength, thanks
```suggestion
  - **Composability**: RaySGD is built on top of the Ray Actor API, enabling seamless integration with existing Ray applications such as RLlib, Tune, and Ray.Serve.
```

File: doc/source/raysgd/raysgd.rst [hunk 12]
```diff
@@ -1,24 +1,65 @@
...
+  - **Scale up and down**: Start on single CPU. Scale up to multi-node, multi-gpu by changing 2 lines of code.
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/raysgd/raysgd.rst]:
```suggestion
  - **Scale up and down**: Start on single CPU. Scale up to multi-node, multi-CPU or multi-GPU clusters by changing 2 lines of code.
```

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 13]
```diff
@@ -3,114 +3,390 @@ RaySGD Pytorch
 
 .. warning:: This is still an experimental API and is subject to change in the near future.
 
-.. tip:: Help us make RaySGD better; take this 1 minute `User Survey <https://forms.gle/26EMwdahdgm7Lscy9>`_!
+.. tip:: Get in touch with us if you're using/considering using `RaySGD <https://forms.gle/26EMwdahdgm7Lscy9>`_!
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]:
```suggestion
.. tip:: Get in touch with us if you're using or considering using `RaySGD <https://forms.gle/26EMwdahdgm7Lscy9>`_!
```

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 14]
```diff
@@ -3,114 +3,390 @@ RaySGD Pytorch
...
 
-Ray's ``PyTorchTrainer`` simplifies distributed model training for PyTorch. The ``PyTorchTrainer`` is a wrapper around ``torch.distributed.launch`` with a Python API to easily incorporate distributed training into a larger Python application, as opposed to needing to execute training outside of Python.
+.. image:: raysgd-pytorch.svg
+    :align: center
 
-----------
+The RaySGD ``PyTorchTrainer`` simplifies distributed model training for PyTorch. The ``PyTorchTrainer`` is a wrapper around ``torch.distributed.launch`` with a Python API to easily incorporate distributed training into a larger Python application, as opposed to needing to execute training outside of Python. For end to end examples, see :ref:`raysgd-pytorch-example`.
 
-**With Ray**:
+Under the hood, ``PytorchTrainer`` will create *replicas* of your model (controlled by ``num_replicas``) which are each managed by a Ray actor.
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]:
```suggestion
Under the hood, ``PytorchTrainer`` will create *replicas* of your model (controlled by ``num_replicas``), each of which is managed by a Ray actor.
```

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 15]
```diff
@@ -3,114 +3,390 @@ RaySGD Pytorch
...
+    def data_creator(config):
+        """Constructs torch.utils.data.Dataset objects.
+
+        Note that even though two Dataset objects can be returned,
+        only one dataset will be used for training.
+
+        Args:
+            config: Configuration dictionary passed into ``PyTorchTrainer``
+
+        Returns:
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]: Good question, I'm thinking about the case where users don't actually care to call ``validate`` (which I find myself doing quite often).  One possibility is to allow users to pass in a data creator in validate, but I can do that in a separate PR.

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 16]
```diff
@@ -3,114 +3,390 @@ RaySGD Pytorch
...
-        backend="nccl")
+        loss_creator=nn.MSELoss,
+        config={"lr": 0.001})
 
-    stats = trainer1.train()
-    print(stats)
-    trainer1.shutdown()
-    print("success!")
 
+You can also set the number of workers and whether the workers are using GPU:
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]:
```suggestion
You can also set the number of workers and whether the workers will use GPUs:
```

File: doc/source/raysgd/raysgd_pytorch.rst [hunk 17]
```diff
@@ -3,114 +3,390 @@ RaySGD Pytorch
...
+Shutting down training
+----------------------
+
+After training, you may want to reappropriate the Ray cluster. To release Ray resources obtained by the trainer:
+
+.. code-block:: python
+
+    trainer.shutdown()
+
+.. note:: Be sure to call ``save`` or ``get_model`` before shutting down.
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]:
```suggestion
.. note:: Be sure to call ``trainer.save()`` or ``trainer.get_model()`` before shutting down.
```

File: doc/source/raysgd/raysgd_pytorch.rst (hunk 7 above)
Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]: Well, I guess in python the return value is always a list/iterable if there's more than 1 returned. 

I think the general case will be to return 1 item, so forcing a list return is quite cumbersome. Internally, it is converted to a list though.

Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]:
```
In [1]: def test():
   ...:     return 1, 2, 3
   ...:

In [2]: a = test()

In [3]: a
Out[3]: (1, 2, 3)

Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]: To be clear this is a valid point - we can open a separate issue/get user feedback down the road.

File: doc/source/raysgd/raysgd_pytorch.rst (hunk 8 above)
Comment (richardliaw) [reply doc/source/raysgd/raysgd_pytorch.rst]:
```suggestion
``PyTorchTrainer`` allows you to run a custom training and validation step in parallel on each worker, providing a flexibility similar to using PyTorch natively. This is done via the ``train_function`` and ``validation_function`` parameters.
```

**Review by simon-mo — APPROVED** [review]

**Comment by richardliaw** [comment]
jenkins test tune

**Comment by richardliaw** [comment]
jenkins test tune

**Comment by richardliaw** [comment]
jenkins test tune
